### ISSUE TITLE
Creation of non expected devices when setting up thermostat types setpoints 

### ISSUE THREAD

**Issue by mealgi** [domoticz/domoticz#5328]
Hello,

Context of the issue :

Version: 2022.1 (build 14338)
Build Hash: 8d73c42d6-modified
Compile Date: 2022-05-15 19:47:20
dzVents Version: 3.1.8
Python Version: 3.7.3 (default, Jan 22 2021, 20:04:44) [GCC 8.3.0]

Issue :
I have developed a custom plugin to support Daikin Altherma Heat Pump system featuring a thermostat (Unit 16).
When setting up a "thermostat" type (242) set point from the Dashboard (namely  ID=00150016 Unit=16 Type=thermostat), domoticz creates a new device ID=0150016 Unit=1 Type=thermostat for this hardware. It creates confusion with the  actual device supported by Unit 1 (ID=00150001 Unit=1 Type=light/switch).

The issue comes from an ID discrepancy between the creation of devices during the onStart function in the python plugin and the one used during the SetSetPointInt function in mainworker.cpp.

In the onStart function the ID name allocated is 00150016 (2 digits for the first ID id1=00 ) whereas in SetSetPoint function in mainworker.cpp it is 0150016 (a one digit for the first ID id1=0).

Proposed solution:
In mainworker.cpp line 10063 in method decode_Thermostat replace
	sprintf(szTmp, "%X%02X%02X%02X", pMeter->id1, pMeter->id2, pMeter->id3, pMeter->id4);
by
	sprintf(szTmp, "%02X%02X%02X%02X", pMeter->id1, pMeter->id2, pMeter->id3, pMeter->id4);

in mainworker.cpp line 12667 in method SetSetPointInt replace
	tmeter.dunit = 1;
by
	tmeter.dunit = Unit;

It is tested successfully in my environment.
BR

**Comment by turin400** [comment]
This isue was discussed in #5229. I think the same thing happens to you.

All the best.

**Comment by mealgi** [comment]
Hello,
I had a look at #5229 and confirm it is the same issue.
I also had a look at the latest version of mainworker.cpp and confirm that my proposed solution still applies to ensure the consistency of thermostat setpoint naming convention (%02X%02X%02X%02X) and use of the correct Unit instead of 1.
BR

**Comment by gizmocuz** [comment]
@mealgi , so it is solved by using the new python method or in a later beta version then you are currently using.

I applied the change in 'SetSetPointInt' , but it is not possible to apply the change for the ID generation.
I have looked in the code of 2019 and this was also generated like this.
The problem is that everyone existing user will get new devices

As the ID generation happens in '**the onStart function**' it needs to be solved/patched there
